### PR TITLE
Fix lack of reset when `iter()` is called on the DALI framework iterator

### PR DESCRIPTION
**Pull request by JanuszL**
- the DALI framework iterator when `iter()` is called checks if any
  data has been consumed yet. If it hasn't it doesn't reset to prevent
  improper operation when DALI FW iterator prefetches the first batch after
  creation, and invocation of something like `enumerate(iterator)`,
  would reset the iterator and extend the length of the first epoch
  (prefetched data in the FW iterator would be added to the freshly
  reset iterator). The flag that is set when any data has been consumed
  was set only in the `next()` method and not `__next__()`. This PR
  fixes this problem and adjusts the test

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Category:
**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
- the DALI framework iterator when `iter()` is called checks if any
  data has been consumed yet. If it hasn't it doesn't reset to prevent
  improper operation when DALI FW iterator prefetches the first batch after
  creation, and invocation of something like `enumerate(iterator)`,
  would reset the iterator and extend the length of the first epoch
  (prefetched data in the FW iterator would be added to the freshly
  reset iterator). The flag that is set when any data has been consumed
  was set only in the `next()` method and not `__next__()`. This PR
  fixes this problem and adjusts the test

### Tests:

- [x] Existing tests apply
  - test_fw_iterators.py:test_*_autoreset_iter()
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

File: dali/test/python/test_fw_iterators.py [hunk 1]
```diff
@@ -1728,7 +1728,9 @@ def BoringPipeline():
 
     loader = fw_iterator(pipeline, reader_name="reader", auto_reset=auto_reset_op)
     for _ in range(2):
-        for i, data in enumerate(loader):
+        loader = iter(loader)
+        for i in range(len(loader)):
+            data = next(loader)
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_fw_iterators.py:1733]: Auotreset kicked in and not call to `iter()`, as enumerate calls `iter()` and handles `StopIteration`.

File: dali/test/python/test_fw_iterators.py [hunk 2]
```diff
@@ -1728,7 +1728,9 @@ def BoringPipeline():
 
     loader = fw_iterator(pipeline, reader_name="reader", auto_reset=auto_reset_op)
     for _ in range(2):
-        for i, data in enumerate(loader):
+        loader = iter(loader)
```
Comment (mzient) [inline dali/test/python/test_fw_iterators.py:1731]: Hmmm... at this point I'd suggest to store it into a separate variable, should we switch to returning something else than `self` in the future.

Comment (JanuszL) [reply dali/test/python/test_fw_iterators.py]: Fixed

**Review by mzient — APPROVED** [review]

**Pull request merged**